### PR TITLE
Make Drools ANC optional but still enabled by default

### PR DESCRIPTION
**Pull request by triceo**
Signed-off-by: Lukáš Petrovický <lpetrovi@redhat.com>

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

We do "simple" maven builds, they are just basically maven commands, but just because we have multiple repositories related between them and one change could affect several of those projects by multiple pull requests, we use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.

[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is not only a github-action tool but a CLI one, so in case you posted multiple pull requests related with this change you can easily reproduce the same build by executing it locally. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] tests</b>
* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] native</b>
</details>

### CODE REVIEW

**Comment by triceo** [comment]
Jenkins run native tests

**Comment by triceo** [comment]
Jenkins run native tests.

**Comment by triceo** [comment]
jenkins run tests
jenkins run native tests

**Comment by triceo** [comment]
jenkins retest this

**Comment by triceo** [comment]
jenkins run native tests

**Comment by triceo** [comment]
@Christopher-Chianelli Can you please quickly check the Gizmo part of this? It's your code, I want to be sure I'm not missing anything.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseBuilder.java
Comment (rsynek) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseBuilder.java:43]: Is this a defensive copy or is the `KieBaseConfiguration` immutable and does not provide any means of creating a copy of it (I am aware it was like that before this PR)?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseBuilder.java
Comment (rsynek) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseBuilder.java:21]: There is no `KieBaseExtractor` in the code base.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsAncConstraintStreamScoreDirectorFactory.java
Comment (rsynek) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsAncConstraintStreamScoreDirectorFactory.java:47]: Nitpick: this method and `DroolsConstraintStreamScoreDirectorFactory.buildKieBase` could be protected to avoid unwanted access from outside the class hierarchy.

**Review by rsynek — COMMENTED** [review]
Please have a look at the method complexity, otherwise looks very good.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsAncConstraintStreamScoreDirectorFactory.java
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsAncConstraintStreamScoreDirectorFactory.java:47]: Unfortunately not, because Quarkus is calling them from Gizmo. (Which is why you see it as unused; Gizmo uses the name of the method as a string, not an actual Java method reference.)

**Review by rsynek — APPROVED** [review]

**Review by Christopher-Chianelli — APPROVED** [review]
Code looks good, but I don't see a test for plain `DroolsConstraintStreamScoreDirectorFactory` is returned when ANC is disabled.

**Comment by triceo** [comment]
On second thought, I decided to not merge this after all.
First of all, I just realized that ANC is not enabled for DRL in Quarkus, even though it could be.
Second, this sort of separation between ANC and non-ANC is inconsistent; sometime it happens inside, sometimes it needs to happen on the outside.
Finally, this change is no longer necessary as the ANC issues have been fixed in another way.

I may get back to this at a later date, but for now, I am closing this.

**Pull request closed without merging**
